crypto/evp/e_chacha20_poly1305.c

```diff
@@ -193,23 +195,158 @@ static int chacha20_poly1305_init_key(EVP_CIPHER_CTX *ctx,
     return 1;
 }
 
+#  if !defined(OPENSSL_SMALL_FOOTPRINT)
+static const unsigned char zero[2 * CHACHA_BLK_SIZE] = { 0 };
+
+static int chacha20_poly1305_tls_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                                        const unsigned char *in, size_t len)
+{
+    EVP_CHACHA_AEAD_CTX *actx = aead_data(ctx);
+    size_t i, tail, tohash_len, plen = actx->tls_payload_length;
+    unsigned char *buf, *tohash, *ctr, storage[2 * CHACHA_BLK_SIZE + 32];
+
+    if (len != plen + POLY1305_BLOCK_SIZE)
+        return -1;
+
+    buf = storage + ((0 - (size_t)storage) & 15);   /* align */
+    ctr = buf + CHACHA_BLK_SIZE;
+    tohash = buf + CHACHA_BLK_SIZE - POLY1305_BLOCK_SIZE;
+
+    if (plen <= CHACHA_BLK_SIZE) {
+        actx->key.counter[0] = 0;
+        ChaCha20_ctr32(buf, zero, 2 * CHACHA_BLK_SIZE, actx->key.key.d,
+                       actx->key.counter);
+        Poly1305_Init(POLY1305_ctx(actx), buf);
+        actx->key.partial_len = 0;
+        memcpy(tohash, actx->tls_aad, POLY1305_BLOCK_SIZE);
+        tohash_len = POLY1305_BLOCK_SIZE;
+        actx->len.aad = EVP_AEAD_TLS1_AAD_LEN;
+        actx->len.text = plen;
+
+        if (ctx->encrypt) {
+            for (i = 0; i < plen; i++) {
+                out[i] = ctr[i] ^= in[i];
+            }
```

Comment:
Fun part for reference. Processing 64 bytes on x86_64. According to profiler processor spends almost **twice** as much time in this loop as in poly1305_blocks[!]. This means that assembly replacement, vectorized of course, would give further significant boost. Though it's not top consumer. Top consumer is ChaCha20_ssse3 and it's possible to optimize it specifically for 2*CHACHA_BLK_SIZE. In combination with loop replacement it should result in ~40% improvement. And with loop replacement it might be meaningful to implement additional case for plen <= 3*CHACHA_BLK_SIZE.